R/getQuote.R

```diff
@@ -24,11 +24,46 @@ function(Symbols,src='yahoo',what, ...) {
   df[Symbols,]
 }
 
+.yahooSession <- function(force.new = FALSE) {
+  cache.name <- "_yahoo_curl_session_"
+  ses <- get0(cache.name, .quantmodEnv) #get cached session
+  
+  if (is.null(ses) || force.new) {
+    ses <- list()
+    ses$h <- curl::new_handle()
+    #yahoo finance doesn't seem to set cookies without these headers 
+    #and the cookies are needed to get the crumb
+    curl::handle_setheaders(ses$h, accept = "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7")
+    URL <- "https://finance.yahoo.com/"
+    r <- curl::curl_fetch_memory(URL, handle = ses$h)
+    #yahoo redirects to a consent form w/ a single cookie for GDPR:
+    #detecting the redirect seems very brittle as its sensitive to the trailing "/"
+    ses$can.crumb <- ((r$status_code == 200) && (URL == r$url) && (NROW(curl::handle_cookies(ses$h)) > 1))
+    assign(cache.name, ses, .quantmodEnv) #cache session
+  }
+
+  if (ses$can.crumb) {
+    #get a crumb so that downstream callers dont have to handle invalid sessions.
+    #this is a network hop, but very lightweight payload
+    n <- if (unclass(Sys.time()) %% 1L >= 0.5) 1L else 2L
+    query.srv <- paste0("https://query", n, ".finance.yahoo.com/v1/test/getcrumb")
+    r <- curl::curl_fetch_memory(query.srv, handle = ses$h)
+    if ((r$status_code == 200) && (length(r$content) > 0)) {
+      ses$crumb = rawToChar(r$content)
+    } else {
+      if (!force.new) ses <- .yahooSession(TRUE) else stop("Unable to get yahoo crumb")
```

Comment:
This seems backward. I expect to call `.yahooSession(TRUE)` when `force.new` is `TRUE`, else throw an error because we can't get a crumb.

Reply:
its really a state machine. the `force.new` is never intended to be used directly, only in the recursive call. i can rewrite it as a loop to clean up the signature

Reply:
Calling the function recursively is fine. The logic on this line (51) confused me, but seems correct once I walk through it.

Assume `force.new` is `FALSE`. Line 45 checks that we have an active session. Line 51 checks if we're able to get a crumb. If we can't get a crumb, then `!force.new` is `TRUE` and we call `.yahooSession(TRUE)`. So that seems like it does what it should.

I think some comments would help me remember what this is doing.

```r
# we were unable to get a crumb
if (force.new) {
    # we couldn't get a crumb with a new session
    stop("unable to get yahoo crumb")
} else {
    # we tried to re-use a session but couldn't get a crumb
    # try to get a crumb using a new session
    ses <- .yahooSession(TRUE)
}
    
```